huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+func DataSourceTopic() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceTopicRead,
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"topic_urn": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"display_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"enterprise_project_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"tags": {
+				Type:     schema.TypeMap,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"push_policy": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceTopicRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.SmnV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating SMN client: %s", err)
+	}
+
+	allTopics, err := topics.List(client).Extract()
+	if err != nil {
+		return diag.Errorf("unable to list topics: %s ", err)
+	}
+
+	log.Printf("[DEBUG] Retrieved SMN topics: %#v", allTopics)
+
+	filter := map[string]interface{}{
+		"TopicUrn":            d.Get("topic_urn"),
+		"Name":                d.Get("name"),
+		"EnterpriseProjectId": d.Get("enterprise_project_id"),
+		"DisplayName":         d.Get("display_name"),
+	}
+
+	filterTopics, err := utils.FilterSliceWithField(allTopics, filter)
+	if err != nil {
+		return diag.Errorf("filter topics failed: %s", err)
+	}
+
+	if len(filterTopics) < 1 {
+		return diag.Errorf("Your query returned no results. " +
+			"Please change your search criteria and try again.")
+	}
+
+	if len(filterTopics) > 1 {
+		return diag.Errorf("Your query returned more than one result." +
+			" Please try a more specific search criteria")
+	}
+
+	topicGet := filterTopics[0].(topics.TopicGet)
+	d.SetId(topicGet.TopicUrn)
+
+	mErr := multierror.Append(
+		d.Set("region", config.GetRegion(d)),
+		d.Set("topic_urn", topicGet.TopicUrn),
+		d.Set("display_name", topicGet.DisplayName),
+		d.Set("name", topicGet.Name),
+		d.Set("push_policy", topicGet.PushPolicy),
+		d.Set("enterprise_project_id", topicGet.EnterpriseProjectId),
+	)
+
+	// fetch tags
+	tagClient, err := config.SmnV2TagClient(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating SMN tag client: %s", err)
+	}
+	tagClient.MoreHeaders = map[string]string{
+		"X-SMN-RESOURCEID-TYPE": "name",
+	}
+	if resourceTags, err := tags.Get(tagClient, "smn_topic", d.Get("name").(string)).Extract(); err == nil {
+		tagmap := utils.TagsToMap(resourceTags.Tags)
+		mErr = multierror.Append(mErr, d.Set("tags", tagmap))
+	} else {
+		log.Printf("[WARN] fetching tags of SMN topic failed: %s", err)
+	}
```

Comment:
how about this
```
resourceTags, err := tags.Get(tagClient, "smn_topic", topicGet.Name).Extract()
if err != nil {
	log.Printf("[WARN] fetching tags of SMN topic failed: %s", err)
}
tagmap := utils.TagsToMap(resourceTags.Tags)
mErr = multierror.Append(mErr, d.Set("tags", tagmap))
```
If we enter the name, we can use the response value to obtain tags.